lenskit-core/src/main/java/org/lenskit/data/entities/AbstractBeanEntity.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2016 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.data.entities;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.reflect.TypeToken;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Abstract entity implementation that uses bean methods.
+ */
+public abstract class AbstractBeanEntity extends AbstractEntity {
+    private static final ConcurrentHashMap<Class<? extends AbstractBeanEntity>, Pair<AttributeSet,ImmutableList<Method>>> cache =
+            new ConcurrentHashMap<>();
+
+    protected final AttributeSet attributes;
```

Comment:
Crap, did I fail to put comments in as to *why* I did this?

There are two purposes for this.

1. Speed up lookup. I micro-benchmarked hash tables, identity hash tables, and a quick linear search through an array with `==`, and the linear search won for small numbers of attributes, on the order of the number of attributes a typical entity has.
2. Decrease memory use. Since attribute sets are constant and interned, each entity with the same set of attributes shares the storage for the attribute names. This isn't as big a deal for the bean entities, since we're caching those, but this lets me use identical attribute management logic between this and the basic entity (and soon the packed entity storage that I am currently working on).